### PR TITLE
Add Many Sanity Checks to Crowdloans

### PR DESCRIPTION
**Pull request by shawntabrizi**
When creating a crowdloan:

* Check that the end date is before the "last possible win date". i.e. if the crowdloan is for lease periods 1 - 4, ensure the end date is at most the beginning of lease period 2. Once we get to lease period 2, we will never have an auction for lease period 1.
* Check the first slot in a crowdloan is at least the current lease period index. For the same reasons, it does not make sense to create a crowdloan for a lease period in the past.

When contributing to a crowdloan:

* Check that the crowdloan is bidding for a valid lease period, i.e. the lease period has not passed already.
* Check the crowdloan account has not already won, i.e. it has some deposit held in the Leaser.

### CODE REVIEW

File: runtime/common/src/crowdloan.rs [hunk 1]
```diff
@@ -150,6 +150,9 @@ pub trait Config: frame_system::Config {
 		LeasePeriod=Self::BlockNumber,
 	>;
 
+	/// The type representing the leasing system.
+	type Leaser: Leaser<AccountId=Self::AccountId, LeasePeriod=Self::BlockNumber>;
```
Comment (kianenigma) [inline runtime/common/src/crowdloan.rs:154]: FYI `type Leaser` in `auctions.rs` has a unrelated doc.

**Review by kianenigma — APPROVED** [review]

**Review by Az111z — APPROVED** [review]

File: runtime/common/src/crowdloan.rs (hunk 1 above)
Comment (shawntabrizi) [reply runtime/common/src/crowdloan.rs:154]: I fixed the doc in this PR?

**Comment by shawntabrizi** [comment]
cc @jacogr

**Comment by shawntabrizi** [comment]
bot merge

**Comment by shawntabrizi** [comment]
bot merge

File: runtime/common/src/crowdloan.rs [hunk 2]
```diff
@@ -254,14 +245,18 @@ decl_event! {
...
-		/// The first slot needs to at least be less than 3 `max_value`.
-		FirstSlotTooFarInFuture,
-		/// Last slot must be greater than first slot.
-		LastSlotBeforeFirstSlot,
-		/// The last slot cannot be more then 3 slots after the first slot.
-		LastSlotTooFarInFuture,
+		/// The current lease period is more than the first lease period.
+		FirstPeriodInPast,
+		/// The first lease period needs to at least be less than 3 `max_value`.
+		FirstPeriodTooFarInFuture,
```
Comment (xlc) [inline runtime/common/src/crowdloan.rs:251]: This is just overflow?

Comment (shawntabrizi) [reply runtime/common/src/crowdloan.rs:251]: What do you mean here? That this error should be replaced with "Overflow" error?

Comment (xlc) [reply runtime/common/src/crowdloan.rs:251]: Yes. Otherwise the description is confusion and I was looking for some max_value constant in trait Config

**Pull request merged**